qutip/solver.py

```diff
@@ -217,111 +222,91 @@ class Options():
-        s += "Options:\n"
-        s += "-----------\n"
-        s += "atol:              " + str(self.atol) + "\n"
-        s += "rtol:              " + str(self.rtol) + "\n"
-        s += "method:            " + str(self.method) + "\n"
-        s += "order:             " + str(self.order) + "\n"
-        s += "nsteps:            " + str(self.nsteps) + "\n"
-        s += "first_step:        " + str(self.first_step) + "\n"
-        s += "min_step:          " + str(self.min_step) + "\n"
-        s += "max_step:          " + str(self.max_step) + "\n"
-        s += "tidy:              " + str(self.tidy) + "\n"
-        s += "num_cpus:          " + str(self.num_cpus) + "\n"
-        s += "norm_tol:          " + str(self.norm_tol) + "\n"
-        s += "norm_steps:        " + str(self.norm_steps) + "\n"
-        s += "rhs_filename:      " + str(self.rhs_filename) + "\n"
-        s += "rhs_reuse:         " + str(self.rhs_reuse) + "\n"
-        s += "seeds:             " + str(seed_length) + "\n"
-        s += "rhs_with_state:    " + str(self.rhs_with_state) + "\n"
-        s += "average_expect:    " + str(self.average_expect) + "\n"
-        s += "average_states:    " + str(self.average_states) + "\n"
-        s += "ntraj:             " + str(self.ntraj) + "\n"
-        s += "store_states:      " + str(self.store_states) + "\n"
-        s += "store_final_state: " + str(self.store_final_state) + "\n"
-
-        return s
+    # Absolute tolerance (default = 1e-8)
+    atol = 1e-8
+    # Relative tolerance (default = 1e-6)
+    rtol = 1e-6
+    # Integration method (default = 'adams', for stiff 'bdf')
+    method = 'adams'
+    # Maximum order used by integrator (<=12 for 'adams', <=5 for 'bdf')
+    order = 12
+    # Max. number of internal steps/call
+    nsteps = 1000
+    # Size of initial step (0 = determined by solver)
+    first_step = 0
+    # Max step size (0 = determined by solver)
+    max_step = 0
+    # Minimal step size (0 = determined by solver)
+    min_step = 0
+    # Average expectation values over trajectories (default = True)
+    average_expect = True
+    # average expectation values
+    average_states = False
+    # tidyup Hamiltonian before calculation (default = True)
+    tidy = True
+    # Number of processors to use (mcsolve only)
+    num_cpus = qset.num_cpus
+    # Tolerance for wavefunction norm (mcsolve only)
+    norm_tol = 1e-3
+    # Tolerance for collapse time precision (mcsolve only)
+    norm_t_tol = 1e-6
+    # Max. number of steps taken to find wavefunction norm to within
+    # norm_tol (mcsolve only)
+    norm_steps = 5
+    # Use preexisting RHS function for time-dependent solvers
+    rhs_reuse = False
+
+    # Use filename for preexisting RHS function (will default to last
+    # compiled function if None & rhs_exists=True)
+    @property
+    def rhs_filename(self):
+        return self._rhs_filename
+
+    @rhs_filename.setter
+    def rhs_filename(self, val):
+        self._rhs_filename = val
```

Comment:
Half. Yes, I use properties as a lazy way to go around the `None`. But adding type hint would not allow to save `None` with rc files so it would not be sufficient by itself. If we stored the settings in a python file instead or save all values as pickled outputs, we would not need to think about types at all. 
But I also want the property also for the read-only install flags such as `has_mkl`. And maybe limit the values some key can take such as `num_cpus`.

Reply:
Yeah, that's fair.  If the parsing's getting too complicated (like full type hints), it's probably more "Python" to write a warning in the documentation, then let the user shoot themselves in the foot if they choose to.